File: backend/app/models/security.py
"""Security assessment models for the Defender Agent."""

from enum import Enum
from pydantic import BaseModel, Field


class Decision(str, Enum):
    """Security decision enum."""

    ALLOW = "ALLOW"
    SANITIZE = "SANITIZE"
    BLOCK = "BLOCK"


class AttackType(str, Enum):
    """Attack type enum."""

    NONE = "NONE"
    PROMPT_INJECTION = "PROMPT_INJECTION"
    OTHER = "OTHER"


class SecurityAssessment(BaseModel):
    """Structured security assessment output from the Defender Agent."""

    decision: Decision = Field(
        ...,
        description="Security decision: ALLOW, SANITIZE, or BLOCK",
    )
    risk_score: int = Field(
        ...,
        ge=0,
        le=100,
        description="Risk score from 0 (safe) to 100 (critical)",
    )
    attack_type: AttackType = Field(
        ...,
        description="Type of attack detected",
    )
    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
        description="Confidence in the assessment from 0.0 to 1.0",
    )
    reason: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Short explanation for the assessment",
    )

File: frontend/frontend/src/types/security.ts
/**
 * TypeScript interfaces for AIShield security API responses.
 * These must match the backend Pydantic models.
 */

export interface Decision {
  ALLOW: 'ALLOW';
  SANITIZE: 'SANITIZE';
  BLOCK: 'BLOCK';
}

/**
 * Attack types as defined in the backend.
 */
export type AttackType = 'NONE' | 'PROMPT_INJECTION' | 'OTHER';

/**
 * Request model for security analysis.
 */
export interface SecurityAnalysisRequest {
  prompt: string;
}

/**
 * Response model for security analysis (POST /analyze and GET /assessments/{id}).
 */
export interface SecurityAnalysisResponse {
  id: number;
  timestamp: string; // ISO datetime string
  prompt: string;
  decision: 'ALLOW' | 'SANITIZE' | 'BLOCK';
  final_risk_score: number; // 0-100
  rule_score: number; // 0-100
  ml_probability: number; // 0.0-1.0
  gemini_risk_score: number; // 0-100
  attack_type: AttackType;
  confidence: number; // 0.0-1.0
  reason: string;
}

/**
 * Individual assessment item for lists (GET /assessments).
 */
export interface AssessmentItem {
  id: number;
  timestamp: string; // ISO datetime string
  prompt: string;
  decision: 'ALLOW' | 'SANITIZE' | 'BLOCK';
  final_risk_score: number; // 0-100
  attack_type: AttackType;
  confidence: number; // 0.0-1.0
  reason: string;
}

/**
 * Response model for paginated assessments (GET /assessments).
 */
export interface AssessmentsResponse {
  items: AssessmentItem[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Response model for security statistics (GET /statistics).
 */
export interface StatisticsResponse {
  total_assessments: number;
  decisions: {
    ALLOW: number;
    SANITIZE: number;
    BLOCK: number;
  };
  average_risk_score: number;
  high_risk_count: number;
  attack_types: {
    NONE: number;
    PROMPT_INJECTION: number;
    OTHER: number;
  };
}

/**
 * Attack analytics item (GET /attacks).
 */
export interface AttackAnalyticsItem {
  attack_type: AttackType;
  count: number;
  percentage: number;
}

/**
 * Risk distribution item (GET /risk-distribution).
 */
export interface RiskDistributionItem {
  range: string; // e.g., "0-19", "20-39", etc.
  count: number;
}

/**
 * Timeline item for assessments over time (GET /timeline).
 */
export interface TimelineItem {
  date: string; // YYYY-MM-DD format
  total: number;
  allowed: number;
  sanitized: number;
  blocked: number;
}
